src/backup.py

```diff
@@ -0,0 +1,86 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provides backup functionality for Synapse."""
+
+import logging
+from typing import Any, Optional
+
+import boto3
+from botocore.exceptions import BotoCoreError, ClientError
+from botocore.exceptions import ConnectionError as BotoConnectionError
+from pydantic import BaseModel, Field, validator
+
+logger = logging.getLogger(__name__)
+
+
+class S3Parameters(BaseModel):
+    """Configuration for accessing S3 bucket.
+
+    Attributes:
+        access_key: AWS access key.
+        secret_key: AWS secret key.
+        region: The region to connect to the object storage.
+        bucket: The bucket name.
+        endpoint: The endpoint used to connect to the object storage.
+        path: The path inside the bucket to store objects.
+        s3_uri_style: The S3 protocol specific bucket path lookup type.
+    """
+
+    access_key: str = Field(alias="access-key")
+    secret_key: str = Field(alias="secret-key")
+    region: Optional[str] = Field(default=None)
+    bucket: str
+    endpoint: Optional[str]
+    path: str = Field(default="")
+    s3_uri_style: str = Field(alias="s3-uri-style", default="host")
+
+    @validator("region", always=True)
+    @classmethod
+    def check_region_or_endpoint_set(cls, region: str, values: dict[str, Any]) -> str:
+        """Validate that either that region or endpoint is set.
+
+        Args:
+            region: region attribute
+            values: all attributes in S3 configuration
+
+        Returns:
+            value of the region attribute
+
+        Raises:
+            ValueError: if the configuration is invalid.
+        """
+        endpoint = values.get("endpoint")
+        if not region and not endpoint:
+            raise ValueError('one of "region" or "endpoint" needs to be set')
+        return region
+
+
+def can_use_bucket(s3_parameters: S3Parameters) -> bool:
```

Comment:
@gregory-schiano Out of the context of this PR: Do you think that charm integrators could provide a function to validate what they are providing? SMTP, SAML, this one S3...Otherwise, all the charms would need something like this before using the configuration the Integrator provides.

Reply:
Shouldn't they fill the relation data only if the service they provide can reasonably be validated ?

Reply:
If your question is, should I trust that integrator in validating that the provided information are effectively correct and working ?
Then I'd say indeed these data should be checked when configured on the integrator and the integrator be blocked if they're not.
But effectively I don't think these validations are in place.
That's a good suggestion for improvement

Reply:
Thanks, I'll close this conversation to not block the PR but let me know if there is some place to add suggestions to the architecture guide or if this could be a story.